src/mongo_odm/nvp.hpp

```diff
@@ -74,17 +58,646 @@ struct Nvp {
+     * @return a SortExpr that reprsents the sort expression {field: +/-1}.
+     */
+    constexpr SortExpr<NvpT> sort(bool ascending) const {
+        return {*static_cast<const NvpT*>(this), ascending};
+    }
+
+    /**
+     * Creates an expression that checks whether the value of this field matches any value in the
+     * given iterable.
+     * @tparam Iterable A type that works in range-based for loops, and yields objects convertible
+     * to the type of this name-value pair.
+     */
+    template <typename Iterable, typename = enable_if_matching_iterable_t<Iterable>>
+    constexpr ComparisonExpr<NvpT, Iterable> in(const Iterable& iter) const {
+        return {*static_cast<const NvpT*>(this), iter, "$in"};
+    }
+
+    /**
+     * Creates an expression that checks whether the value of this field matches none of the values
+     * in the given iterable.
+     * @tparam Iterable A type that works in range-based for loops, and yields objects convertible
+     * to the type of this name-value pair.
+     */
+    template <typename Iterable, typename = enable_if_matching_iterable_t<Iterable>>
+    constexpr ComparisonExpr<NvpT, Iterable> nin(const Iterable& iter) const {
+        return {*static_cast<const NvpT*>(this), iter, "$nin"};
+    }
+
+    /**
+     * Creates an expression that checks the existence of a certain field.
+     * This is only enabled for fields that are optional types.
+     * {field: {$exists: <bool>}}
+     * @param exists    If false, checks that the given field does *not* exist in a document.
+     *                  True by default.
+     */
+    template <typename U = T, typename = typename std::enable_if<is_optional<U>::value>::type>
+    constexpr ComparisonExpr<NvpT, bool> exists(const bool& exists) const {
+        return {*static_cast<const NvpT*>(this), exists, "$exists"};
+    }
+
+    /**
+     * Creates a ModExpr that represents a query with the $mod operator.
+     * Such a query essentially checks that "nvp_value % divisor == remainder"
+     * @param divisor   The divisor for the modulus operation
+     * @param remainder   The remainder after dividing a value by `divisor`
+     * @returns A ModExpr representing this query.
+     */
+    template <typename U = no_opt_type,
+              typename = typename std::enable_if<std::is_arithmetic<U>::value>::type>
+    constexpr ModExpr<NvpT> mod(const int& divisor, const int& remainder) const {
+        return {*static_cast<const NvpT*>(this), divisor, remainder};
+    }
+
+    /**
+     * Creates a comparison expression that represents a query with a $regex operator.
+     * Such a query only works for string fields.
+     * @param regex     The regex to check against.
+     * @param options   Options to pass to the regex.
+     */
+    template <typename U = no_opt_type,
+              typename = typename std::enable_if<is_string<U>::value>::type>
+    constexpr ComparisonValueExpr<NvpT, bsoncxx::types::b_regex> regex(const char* regex,
```

Comment:
ah, did not know that. 

The $regex operator can accept either, but the $not operator is more finnicky. The main reason I'm using b_regex here is that the `$not` operator can't have a `$regex` operator inside of it, so when a user applies `operator!` to a regex expression, it would transform it to a `$not` that directly contains the regex object. 

i.e. `{"z" : {$not: /O+D/i }}` works find in the mongo shell, but `{"z" : {$not: {$regex: /O+D/i} }}` produces an error: `"errmsg": "$not cannot have a regex"`